HW3/P5/label_regions.cl

```diff
@@ -79,21 +82,53 @@ propagate_labels(__global __read_write int *labels,
     // the pixel for this thread
     old_label = buffer[buf_y * buf_w + buf_x];
 
+    int this_label;
+    int last_label = -1;
+    int last_index = -1;
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+    if ((lx == 0) && (ly == 0)) {
+        for (int bx = halo; bx < lsize_x + halo; bx++) {
+            for (int by = halo; by < lsize_y + halo; by++) {
+                this_label = buffer[bx + buf_w * by];
+                if (this_label < w*h) {
+                    if (this_label != last_label) {
+                        buffer[bx + buf_w * by] = labels[this_label];
+                        last_label = this_label;
+                    }
+                    else {
+                        buffer[bx + buf_w * by] = buffer[last_index];
+                    }
+                    last_index = bx + buf_w * by;
+                }
+            }
+        }
+    }
+
+    // Make sure local buffer values have finished updating
+    barrier(CLK_LOCAL_MEM_FENCE);
     
     // stay in bounds
-    if ((x < w) && (y < h)) {
+    if ((x < w) && (y < h) && (old_label < w*h)) {
         // CODE FOR PART 1 HERE
-        // We set new_label to the value of old_label, but you will need
-        // to adjust this for correctness.
         new_label = old_label;
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.